### PR TITLE
Allow for all possible kinds of references

### PR DESCRIPTION
**Pull request by kim**
This entails:

* an `ActiveSpan` is mutable, thus derefencing it as a `ChildOf`
  reference must occur at the end of it's lifetime (ie. we have to carry
  it around as a mutable reference)
* `traced` returns the `FinishedSpan` along with the result of the
  traced action, such that it can be used as a `FollowsFrom` reference
  in subsequent `traced` calls
* additionally, pure `Reference`s may be used as obtained from the
  propagation carrier

### CODE REVIEW

File: opentracing/OpenTracing.hs [hunk 1]
```diff
@@ -14,11 +14,17 @@ module OpenTracing
     , traced
     , traced'
 
+    , traced_
+    , traced__
+    , traced'_
+    , traced'__
```
Comment (isovector) [inline opentracing/OpenTracing.hs]: these are some aggressive synonyms :)

**Comment by isovector** [comment]
I'm not entirely sure what the value of FollowsFrom is, but besides that everything looks good!

File: opentracing/OpenTracing.hs (hunk 1 above)
Comment (kim) [reply opentracing/OpenTracing.hs]: Haha, I figured the RHSes would get old very quickly ;)

**Comment by kim** [comment]
> I'm not entirely sure what the value of FollowsFrom is

I think it's when you don't want to wait for the parent span to complete until all children have completed (see also: https://github.com/opentracing/opentracing-go/blob/master/tracer.go#L183-L207). Contrived example:

```haskell
users <- traced (spanOpts "getUsers" mempty) $ const getUsers
cache <- async $ traced (spanOpts "cacheUserIds" (followsFrom (tracedSpan users))) $ const cacheUserIds (tracedResult users)
doSomethingElseWith (tracedResult users)
wait cache
```

If the cache insert happens as a child, the `users` result is only available after that action has completed (and includes it's timing information).

Obviously, nothing prevents you at the moment to run children asynchronously and _don't_ wait for their completion inside the `traced` block. I think the resulting trace would look the same.

**Pull request merged**